### PR TITLE
Changes to run the recognition (acceptwaveform) on a different thread

### PR DESCRIPTION
**Pull request by jose-alvessh**

### CODE REVIEW

File: android/lib/src/main/java/org/vosk/android/SpeechService.java [hunk 1]
```diff
@@ -34,14 +40,20 @@ public class SpeechService {
...
     private final int sampleRate;
-    private final static float BUFFER_SIZE_SECONDS = 0.2f;
+    private final static float BUFFER_SIZE_SECONDS = 0.4f;
+    private final static int RECOGNIZER_QUEUE = 5;
     private final int bufferSize;
     private final AudioRecord recorder;
 
     private RecognizerThread recognizerThread;
 
+    private boolean accepted = false;
```
Comment (pedhensan) [inline android/lib/src/main/java/org/vosk/android/SpeechService.java]: you can clean this variable, we are using a local variable with the same name in the TranscriptionThread

File: android/lib/src/main/java/org/vosk/android/SpeechStreamService.java [hunk 2]
```diff
@@ -126,16 +131,29 @@ public RecognizerThread(RecognitionListener listener) {
...
 
             byte[] buffer = new byte[bufferSize];
+            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
 
             while (!interrupted()
                     && ((timeoutSamples == NO_TIMEOUT) || (remainingSamples > 0))) {
                 try {
                     int nread = inputStream.read(buffer, 0, buffer.length);
+
+                    byte[] bdata = new byte[nread * 2];
```
Comment (pedhensan) [inline android/lib/src/main/java/org/vosk/android/SpeechStreamService.java]: This is old code that can be removed - we are not outputting the resultBuffer, right?

File: android/lib/src/main/java/org/vosk/android/SpeechService.java [hunk 3]
```diff
@@ -227,31 +253,91 @@ public void run() {
...
 
-                if (recognizer.acceptWaveForm(buffer, nread)) {
-                    final String result = recognizer.getResult();
-                    mainHandler.post(() -> listener.onResult(result));
-                } else {
-                    final String partialResult = recognizer.getPartialResult();
-                    mainHandler.post(() -> listener.onPartialResult(partialResult));
+                // Adds the data read from the AudioRecorder to the queue of chunks of audio
+                try {
+                    transcriptionThread.recordingChunksQueue.put(new RecordingChunk(buffer, nread));
```
Comment (pedhensan) [inline android/lib/src/main/java/org/vosk/android/SpeechService.java:266]: Just checking that we are doing this on purpose
* queue.offer() is a synchronous attempt at add the element, which fails if the queue is full
* queue.put() is a synchronous attempt to add the element which will block until the queue is no longer full

I think the better idea would be to:
* Check the queue remaining capacity
* If it has a couple of elements lagging output that we are lagging (a method onTranscriptionDelayed() or something like that)
* If it is full output that we are losing data (onDataProcessingFailed() or something like that)
* Then use the put to have a blocking insert

File: android/lib/src/main/java/org/vosk/android/SpeechService.java [hunk 4]
```diff
@@ -227,31 +253,91 @@ public void run() {
...
+        private final int nread;
+        private final short[] audioBuffer;
 
-            if (!paused) {
-                // If we met timeout signal that speech ended
-                if (timeoutSamples != NO_TIMEOUT && remainingSamples <= 0) {
-                    mainHandler.post(() -> listener.onTimeout());
-                } else {
-                    final String finalResult = recognizer.getFinalResult();
-                    mainHandler.post(() -> listener.onFinalResult(finalResult));
```
Comment (pedhensan) [inline android/lib/src/main/java/org/vosk/android/SpeechService.java:251]: we no longer have the onTImeout() and onFinalResult() callbacks, are they not necessary?

File: android/lib/src/main/java/org/vosk/android/RecognitionListener.java [hunk 5]
```diff
@@ -27,7 +27,7 @@ public interface RecognitionListener {
     /**
      * Called after silence occured.
      */
-    void onResult(String hypothesis);
+    public void onResult(String hypothesis);
```
Comment (pedhensan) [inline android/lib/src/main/java/org/vosk/android/RecognitionListener.java]: interface methods are public by default, they do not need the qualifier

File: android/lib/src/main/java/org/vosk/android/SpeechService.java (hunk 4 above)
Comment (jose-alvessh) [reply android/lib/src/main/java/org/vosk/android/SpeechService.java:251]: Well we don't use these callbacks and therefore I'm removing them in order to not get possible overhead running them without being necessary.

Comment (pedhensan) [reply android/lib/src/main/java/org/vosk/android/SpeechService.java:251]: 👍

**Review by pedhensan — APPROVED** [review]
cool 👍

**Pull request merged**